kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+///     - If a `cdc` action was found in the prepare phase, only `cdc` actions are selected
+///     - Otherwise, select `add` and `remove` actions. Note that only `remove` actions that do not
+///       share a path with an `add` action are selected.
+///
+/// Note: As a consequence of the two phases, LogReplayScanner will iterate over each action in the
+/// commit twice. It also may use an unbounded amount of memory, proportional to the number of
+/// `add` and `remove` actions in the commit.
+struct LogReplayScanner {
+    // True if a `cdc` action was found after running [`LogReplayScanner::prepare_phase`]
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    // The commit file that this replay scanner will operate on.
+    commit_file: ParsedLogPath,
+    // The timestamp associated with this commit. By default this is the file modification time
+    // from the commit's [`FileMeta`]. If there is a [`CommitInfo`] with a timestamp generated by
+    // in-commit timestamps, that timestamp will be used instead.
+    timestamp: i64,
+    // The schema of the table
+    schema: SchemaRef,
+    json_handler: Arc<dyn JsonHandler>,
+    filter: Option<DataSkippingFilter>,
+    expression_handler: Arc<dyn ExpressionHandler>,
+}
+
+impl LogReplayScanner {
+    fn new(
+        commit_file: ParsedLogPath,
+        json_handler: Arc<dyn JsonHandler>,
+        filter: Option<DataSkippingFilter>,
+        expression_handler: Arc<dyn ExpressionHandler>,
+        schema: SchemaRef,
+    ) -> Self {
+        Self {
+            timestamp: commit_file.location.last_modified,
+            commit_file,
+            json_handler,
+            filter,
+            has_cdc_action: Default::default(),
+            remove_dvs: Default::default(),
+            expression_handler,
+            schema,
+        }
+    }
+    fn prepare_phase(&mut self) -> DeltaResult<()> {
+        let schema = PreparePhaseVisitor::schema()?;
+        let action_iter = self.json_handler.read_json_files(
+            &[self.commit_file.location.clone()],
+            schema,
+            None,
+        )?;
+
+        let mut add_paths = HashSet::new();
+        for actions in action_iter {
+            let actions = actions?;
+
+            // Apply data skipping to get back a selection vector for actions that passed skipping.
+            // We start our selection vector based on what was filtered. We will add to this vector
+            // below if a file has been removed. Note: None implies all files passed data skipping.
+            let selection_vector = match &self.filter {
+                Some(filter) => filter.apply(actions.as_ref())?,
+                None => vec![true; actions.len()],
+            };
+
+            let mut visitor = PreparePhaseVisitor::new(self, selection_vector, &mut add_paths);
```

Comment:
There's a visitor for every file because we only need to visit a single commit a time. (ie a single json commit file)

For example, `has_cdc` is true only if the commit has a CDC, but we have to check every record batch in the commit.  The visitor doesn't make sense across multiple commits. Hence a single commit/file is the correct granularity for this.